File: web/src/lib/server/bot/redemptions.ts
import type { BlindBoxSeries } from '$lib/contracts/catalog';
import type {
  BlindBoxService,
  Logger,
  OverlayBus,
  Random,
  StatsService,
} from '$lib/server/application/ports';
import { pickWeightedPlushie } from '$lib/server/domain/blind-box/random';
import { formatStats } from '$lib/server/domain/stats/format';

import { randomInteger } from './random';
import type { Redemption, SendChat } from './types';

export interface RedemptionDependencies {
  series: readonly BlindBoxSeries[];
  stats: StatsService;
  blindBox: BlindBoxService;
  overlay: OverlayBus;
  random: Random;
  logger: Logger;
  sendChat: SendChat;
}

const redemptionFailureMessage = (username: string): string =>
  `@${username} sorry, the redemption failed. Please ping @modservo.`;

export const createRedemptions = ({
  series,
  stats,
  blindBox,
  overlay,
  random,
  logger,
  sendChat,
}: RedemptionDependencies): ReadonlyMap<string, Redemption> => {
  const redemptions = new Map<string, Redemption>();

  redemptions.set('Drink a Potion', async (event, signal) => {
    try {
      await stats.getOrCreate(event.userID, event.username);
    } catch (error) {
      logger.error('failed to get or create stats', { error, user: event.username });

      return;
    }

    if (signal.aborted) {
      return;
    }

    let definition;
    try {
      const index = randomInteger(random, stats.definitions.length, 'stat definition');
      definition = stats.definitions[index];
    } catch (error) {
      logger.error('failed to get random stat definition', { error });

      return;
    }

    const lost = randomInteger(random, 100, 'potion outcome') < 5;
    const amount = lost ? -1 : 1;

    try {
      await stats.adjust(event.userID, definition.name, amount);
    } catch (error) {
      logger.error('failed to modify stat', { error, user: event.username });

      return;
    }

    if (signal.aborted) {
      return;
    }

    const outcome = lost ? 'lost' : 'gained';
    await sendChat(
      `A shifty looking merchant hands ${event.username} a glittering potion. Without hesitation, they sink the whole drink. ${event.username} ${outcome} ${definition.longName}`,
      signal,
    );

    if (signal.aborted) {
      return;
    }

    let values;
    try {
      values = await stats.get(event.userID);
    } catch (error) {
      logger.error('failed to get stats', { error, user: event.username });

      return;
    }

    if (signal.aborted) {
      return;
    }

    await sendChat(formatStats(event.username, values), signal);
  });

  redemptions.set('Tempt the Dice', async (event, signal) => {
    await sendChat(`${event.username} has rolled with initiative.`, signal);

    if (signal.aborted) {
      return;
    }

    let values;
    try {
      values = await stats.getOrCreate(event.userID, event.username);
    } catch (error) {
      logger.error('failed to get stats', { error, user: event.username });

      return;
    }

    if (signal.aborted) {
      return;
    }

    await sendChat(formatStats(event.username, values), signal);
  });

  for (const config of series) {
    redemptions.set(config.redemptionTitle, async (event, signal) => {
      let plushie;
      try {
        plushie = pickWeightedPlushie(config.plushies, random);
      } catch (error) {
        logger.error('failed to pick plushie', { error, series: config.series });
        await sendChat(redemptionFailureMessage(event.username), signal);

        return;
      }

      let result;
      try {
        result = await blindBox.grant(event.userID, event.username, config.series, plushie.key);
      } catch (error) {
        logger.error('failed to redeem blind box', { error, user: event.username });
        await sendChat(redemptionFailureMessage(event.username), signal);

        return;
      }

      if (signal.aborted) {
        return;
      }

      overlay.publish({
        type: 'blindbox_redemption',
        username: event.username,
        plushie,
        isNew: result.isNew,
        collection: result.collection,
        config,
      });
      logger.info('blind box redeemed', {
        user: event.username,
        series: config.series,
        plushie: plushie.key,
        isNew: result.isNew,
      });
    });
  }

  return redemptions;
};

File: charsibot/redemptions.go
package charsibot

import (
	"context"
	"fmt"
	"math/rand/v2"

	"github.com/joeyak/go-twitch-eventsub/v3"

	"github.com/lukeramljak/charsibot/blindbox"
	"github.com/lukeramljak/charsibot/server"
	"github.com/lukeramljak/charsibot/stats"
)

// RedemptionFunc handles a channel point redemption event.
type RedemptionFunc func(ctx context.Context, b *Bot, event twitch.EventChannelChannelPointsCustomRewardRedemptionAdd)

// Redemptions returns the full map of channel point redemptions keyed by reward title.
func Redemptions(seriesConfigs []blindbox.SeriesConfig) map[string]RedemptionFunc {
	const rewardDrinkAPotion = "Drink a Potion"

	redemptions := map[string]RedemptionFunc{
		rewardDrinkAPotion: func(ctx context.Context, b *Bot, event twitch.EventChannelChannelPointsCustomRewardRedemptionAdd) {
			const (
				negativePercent = 5
				percentMax      = 100
			)

			userID := event.UserID
			username := event.UserName

			if _, err := b.statsService.GetOrCreateStats(ctx, userID, username); err != nil {
				b.logger.Error("failed to get or create stats", "err", err, "user", username)
				return
			}

			stat, err := b.statsService.GetRandomStatDefinition(ctx)
			if err != nil {
				b.logger.Error("failed to get random stat definition", "err", err)
				return
			}

			delta := int64(1)
			outcome := "gained"
			roll := rand.IntN(percentMax)
			if roll < negativePercent {
				delta = -1
				outcome = "lost"
			}

			if err = b.statsService.ModifyStatValue(ctx, userID, stat.Name, delta); err != nil {
				b.logger.Error("failed to modify stat", "err", err, "user", username)
				return
			}

			b.SendMessage(SendMessageParams{
				Message: "A shifty looking merchant hands " + username + " a glittering potion. Without hesitation, they sink the whole drink. " +
					username + " " + outcome + " " + stat.LongName,
			})

			userStats, err := b.statsService.GetUserStats(ctx, userID)
			if err != nil {
				b.logger.Error("failed to get stats", "err", err, "user", username)
				return
			}
			b.SendMessage(SendMessageParams{Message: stats.FormatStats(username, userStats)})
		},

		"Tempt the Dice": func(ctx context.Context, b *Bot, event twitch.EventChannelChannelPointsCustomRewardRedemptionAdd) {
			userID := event.UserID
			username := event.UserName

			b.SendMessage(SendMessageParams{
				Message: username + " has rolled with initiative.",
			})

			userStats, err := b.statsService.GetOrCreateStats(ctx, userID, username)
			if err != nil {
				b.logger.Error("failed to get stats", "err", err, "user", username)
				return
			}
			b.SendMessage(SendMessageParams{Message: stats.FormatStats(username, userStats)})
		},
	}

	for _, cfg := range seriesConfigs {
		redemptions[cfg.RedemptionTitle] = func(ctx context.Context, b *Bot, event twitch.EventChannelChannelPointsCustomRewardRedemptionAdd) {
			redeemBlindBox(ctx, b, event.UserID, event.UserName, cfg)
		}
	}

	return redemptions
}

// redeemBlindBox picks a random plushie, records it, and broadcasts the SSE event.
func redeemBlindBox(ctx context.Context, b *Bot, userID, username string, cfg blindbox.SeriesConfig) {
	plushie, err := blindbox.PickPlushie(cfg.Plushies)
	if err != nil {
		b.logger.Error("failed to pick plushie", "err", err, "series", cfg.Series)
		b.SendMessage(
			SendMessageParams{
				Message: fmt.Sprintf("@%s sorry, the redemption failed. Please ping @modservo.", username),
			},
		)
		return
	}

	result, err := b.blindboxService.Redeem(ctx, userID, username, cfg.Series, plushie.Key)
	if err != nil {
		b.logger.Error("failed to redeem blind box", "err", err, "user", username)
		b.SendMessage(
			SendMessageParams{
				Message: fmt.Sprintf("@%s sorry, the redemption failed. Please ping @modservo.", username),
			},
		)
		return
	}

	b.broadcast(server.OverlayEvent{
		Type: server.EventTypeBlindBoxRedemption,
		Data: blindbox.BlindBoxRedemptionData{
			Username:   result.Username,
			Plushie:    plushie,
			IsNew:      result.IsNew,
			Collection: result.Collection,
			Config:     cfg,
		},
	})
	b.logger.Info(
		"blind box redeemed",
		"user",
		username,
		"series",
		cfg.Series,
		"plushie",
		plushie.Key,
		"is_new",
		result.IsNew,
	)
}
